contracts/RandomBeaconHistory.cdc

```diff
@@ -36,14 +49,127 @@ access(all) contract RandomBeaconHistory {
+            // This happens in the rare case where a new gap occurs because of a system transaction failure.
+            while nextIndex > UInt64(RandomBeaconHistory.randomSourceHistory.length) {
+                RandomBeaconHistory.randomSourceHistory.append([])
+            }
+
+            // we are now at the correct index to record the source of randomness
+            // created by the protocol for the current block
             RandomBeaconHistory.randomSourceHistory.append(randomSourceHistory)
+
+            // check for any existing gap and backfill using the input random source if needed.
+            backfiller.backfill(randomSource: randomSourceHistory)
+        }
+    }
+
+    /* --- Backfiller --- */
+    //
+    /// A recovery mechanism designed to backfill missed sources of randomness in the event of a missed commitment due
+    /// to a system transaction failure.
+    ///
+    access(all) resource Backfiller {
+        /// Start index of the first gap in the `randomSourceHistory` array where random sources were not recorded, 
+        /// because of a heartbeat failure.
+        /// There may be non contiguous gaps in the history, `gapStartIndex` is the start index of the lowest-height
+        /// gap.
+        /// If no gaps exist, `gapStartIndex` is equal to the `randomSourceHistory` array length.
+        access(contract) var gapStartIndex: UInt64
+
+        init() {
+            self.gapStartIndex = UInt64(RandomBeaconHistory.randomSourceHistory.length)
+        }
+
+        /// Backfills possible empty entries (gaps) in the history array starting from the stored `gapStartIndex`,
+        /// using `randomSource` as a seed for all entries.
+        /// If there are no gaps, `gapStartIndex` is just updated to `RandomBeaconHistory`'s length.
+        //
+        /// When backfilling, all entries use the same entropy. Each entry is extracted from `randomSource` using
+        /// successive hashing. This makes sure the entries are all distinct although they provide
+        /// the same entropy.
+        //
+        /// gaps only occur in the rare event of a system transaction failure. In this case, entries are still
+        /// filled using a source not known at the time of block execution, which guarantees unpredicatability.
+        access(contract) fun backfill(randomSource: [UInt8]) {
+            let arrayLength = UInt64(RandomBeaconHistory.randomSourceHistory.length)
+            // optional optimization for the happy common path: if no gaps are detected, 
+            // backfilling isn't needed, return early
+            if self.gapStartIndex + 1 == arrayLength {
+                self.gapStartIndex = arrayLength
+                return
+            }
+
+            // backFilling is limited to a max entries only to limit the computation cost.
+            // This means a large gap may need a few transactions to get fully backfilled.
+            let maxEntries = 100
```

Comment:
https://github.com/onflow/flow-core-contracts/pull/416/commits/c5b9d92a2fc2e3a4a1bb2de8b33cd3347b3fd990

Please have a careful look at this one since I'm very new to Cadence. In particular, I made `borrowBackfiller` public (not sure if that's fine). I also remember the heartbeat function is only callable from the service account, the setter of the backfiller limit should also be callable only by the service account. Is this still the case?